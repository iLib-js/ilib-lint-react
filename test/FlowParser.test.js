/*
 * FlowParser.test.js - test the React JS + flow parser
 *
 * Copyright Â© 2023 Box, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ResourceString } from 'ilib-tools-common';

import FlowParser from '../src/parsers/FlowParser.js';

import { Result, IntermediateRepresentation } from 'i18nlint-common';

describe("testFlowParser", () => {
    test("FlowParserConstructorEmpty", () => {
        expect.assertions(1);

        const parser = new FlowParser();
        expect(parser).toBeTruthy();
    });

    test("FlowParserConstructorPath", () => {
        expect.assertions(1);

        const parser = new FlowParser({
            filePath: "./test/testfiles/testfile.js"
        });
        expect(parser).toBeTruthy();
    });

    test("FlowParserGetDescription", () => {
        expect.assertions(2);

        const parser = new FlowParser();
        expect(parser).toBeTruthy();

        expect(parser.getDescription()).toBe("A parser for JS and JSX files with flow type definitions.");
    });

    test("FlowParserGetName", () => {
        expect.assertions(2);

        const parser = new FlowParser();
        expect(parser).toBeTruthy();

        expect(parser.getName()).toBe("js");
    });

    test("FlowParserGetExtensions", () => {
        expect.assertions(2);

        const parser = new FlowParser();
        expect(parser).toBeTruthy();

        expect(parser.getExtensions()).toStrictEqual([ "js", "jsx" ]);
    });

    test("FlowParserSimple", () => {
        expect.assertions(3);

        const parser = new FlowParser();
        expect(parser).toBeTruthy();

        const actual = parser.parseString(
            `// @flow
            import foo from '../src/index.js';
            export default function pathToFile(): string {
                return foo('x/y');
            }
            `, "x/y");
        expect(actual).toBeTruthy();
        const actualSimplified = JSON.parse(JSON.stringify(actual));

        expect(actualSimplified).toMatchSnapshot();
    });

    test("FlowParserMoreComplex", () => {
        expect.assertions(3);

        const parser = new FlowParser();
        expect(parser).toBeTruthy();

        const actual = parser.parseString(
            `// comment
            import foo from '../src/index.js';

            const str: string = "String";

            export default str;
            `, "x/y");
        expect(actual).toBeTruthy();
        const actualSimplified = JSON.parse(JSON.stringify(actual));

        expect(actualSimplified).toMatchSnapshot();
    });
});

