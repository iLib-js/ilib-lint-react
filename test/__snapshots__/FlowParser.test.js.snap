// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testFlowParser FlowParserMoreComplex 1`] = `
{
  "filePath": "x/y",
  "ir": {
    "body": [
      {
        "importKind": "value",
        "leadingComments": [
          {
            "loc": {
              "end": {
                "column": 10,
                "line": 1,
              },
              "source": null,
              "start": {
                "column": 0,
                "line": 1,
              },
            },
            "range": [
              0,
              10,
            ],
            "type": "Line",
            "value": " comment",
          },
        ],
        "loc": {
          "end": {
            "column": 46,
            "line": 2,
          },
          "source": null,
          "start": {
            "column": 12,
            "line": 2,
          },
        },
        "range": [
          23,
          57,
        ],
        "source": {
          "loc": {
            "end": {
              "column": 45,
              "line": 2,
            },
            "source": null,
            "start": {
              "column": 28,
              "line": 2,
            },
          },
          "range": [
            39,
            56,
          ],
          "raw": "'../src/index.js'",
          "type": "Literal",
          "value": "../src/index.js",
        },
        "specifiers": [
          {
            "loc": {
              "end": {
                "column": 22,
                "line": 2,
              },
              "source": null,
              "start": {
                "column": 19,
                "line": 2,
              },
            },
            "local": {
              "loc": {
                "end": {
                  "column": 22,
                  "line": 2,
                },
                "source": null,
                "start": {
                  "column": 19,
                  "line": 2,
                },
              },
              "name": "foo",
              "optional": false,
              "range": [
                30,
                33,
              ],
              "type": "Identifier",
              "typeAnnotation": null,
            },
            "range": [
              30,
              33,
            ],
            "type": "ImportDefaultSpecifier",
          },
        ],
        "type": "ImportDeclaration",
      },
      {
        "declarations": [
          {
            "id": {
              "loc": {
                "end": {
                  "column": 29,
                  "line": 4,
                },
                "source": null,
                "start": {
                  "column": 18,
                  "line": 4,
                },
              },
              "name": "str",
              "optional": false,
              "range": [
                77,
                88,
              ],
              "type": "Identifier",
              "typeAnnotation": {
                "loc": {
                  "end": {
                    "column": 29,
                    "line": 4,
                  },
                  "source": null,
                  "start": {
                    "column": 21,
                    "line": 4,
                  },
                },
                "range": [
                  80,
                  88,
                ],
                "type": "TypeAnnotation",
                "typeAnnotation": {
                  "loc": {
                    "end": {
                      "column": 29,
                      "line": 4,
                    },
                    "source": null,
                    "start": {
                      "column": 23,
                      "line": 4,
                    },
                  },
                  "range": [
                    82,
                    88,
                  ],
                  "type": "StringTypeAnnotation",
                },
              },
            },
            "init": {
              "loc": {
                "end": {
                  "column": 40,
                  "line": 4,
                },
                "source": null,
                "start": {
                  "column": 32,
                  "line": 4,
                },
              },
              "range": [
                91,
                99,
              ],
              "raw": ""String"",
              "type": "Literal",
              "value": "String",
            },
            "loc": {
              "end": {
                "column": 40,
                "line": 4,
              },
              "source": null,
              "start": {
                "column": 18,
                "line": 4,
              },
            },
            "range": [
              77,
              99,
            ],
            "type": "VariableDeclarator",
          },
        ],
        "kind": "const",
        "loc": {
          "end": {
            "column": 41,
            "line": 4,
          },
          "source": null,
          "start": {
            "column": 12,
            "line": 4,
          },
        },
        "range": [
          71,
          100,
        ],
        "type": "VariableDeclaration",
      },
      {
        "declaration": {
          "loc": {
            "end": {
              "column": 30,
              "line": 6,
            },
            "source": null,
            "start": {
              "column": 27,
              "line": 6,
            },
          },
          "name": "str",
          "optional": false,
          "range": [
            129,
            132,
          ],
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "exportKind": "value",
        "loc": {
          "end": {
            "column": 31,
            "line": 6,
          },
          "source": null,
          "start": {
            "column": 12,
            "line": 6,
          },
        },
        "range": [
          114,
          133,
        ],
        "type": "ExportDefaultDeclaration",
      },
    ],
    "comments": [
      {
        "loc": {
          "end": {
            "column": 10,
            "line": 1,
          },
          "source": null,
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "range": [
          0,
          10,
        ],
        "type": "Line",
        "value": " comment",
      },
    ],
    "errors": [],
    "loc": {
      "end": {
        "column": 31,
        "line": 6,
      },
      "source": null,
      "start": {
        "column": 12,
        "line": 2,
      },
    },
    "range": [
      23,
      133,
    ],
    "type": "Program",
  },
  "type": "ast-jstree",
}
`;

exports[`testFlowParser FlowParserSimple 1`] = `
{
  "filePath": "x/y",
  "ir": {
    "body": [
      {
        "importKind": "value",
        "loc": {
          "end": {
            "column": 46,
            "line": 2,
          },
          "source": null,
          "start": {
            "column": 12,
            "line": 2,
          },
        },
        "range": [
          21,
          55,
        ],
        "source": {
          "loc": {
            "end": {
              "column": 45,
              "line": 2,
            },
            "source": null,
            "start": {
              "column": 28,
              "line": 2,
            },
          },
          "range": [
            37,
            54,
          ],
          "raw": "'../src/index.js'",
          "type": "Literal",
          "value": "../src/index.js",
        },
        "specifiers": [
          {
            "loc": {
              "end": {
                "column": 22,
                "line": 2,
              },
              "source": null,
              "start": {
                "column": 19,
                "line": 2,
              },
            },
            "local": {
              "loc": {
                "end": {
                  "column": 22,
                  "line": 2,
                },
                "source": null,
                "start": {
                  "column": 19,
                  "line": 2,
                },
              },
              "name": "foo",
              "optional": false,
              "range": [
                28,
                31,
              ],
              "type": "Identifier",
              "typeAnnotation": null,
            },
            "range": [
              28,
              31,
            ],
            "type": "ImportDefaultSpecifier",
          },
        ],
        "type": "ImportDeclaration",
      },
      {
        "declaration": {
          "async": false,
          "body": {
            "body": [
              {
                "argument": {
                  "arguments": [
                    {
                      "loc": {
                        "end": {
                          "column": 32,
                          "line": 4,
                        },
                        "source": null,
                        "start": {
                          "column": 27,
                          "line": 4,
                        },
                      },
                      "range": [
                        142,
                        147,
                      ],
                      "raw": "'x/y'",
                      "type": "Literal",
                      "value": "x/y",
                    },
                  ],
                  "callee": {
                    "loc": {
                      "end": {
                        "column": 26,
                        "line": 4,
                      },
                      "source": null,
                      "start": {
                        "column": 23,
                        "line": 4,
                      },
                    },
                    "name": "foo",
                    "optional": false,
                    "range": [
                      138,
                      141,
                    ],
                    "type": "Identifier",
                    "typeAnnotation": null,
                  },
                  "loc": {
                    "end": {
                      "column": 33,
                      "line": 4,
                    },
                    "source": null,
                    "start": {
                      "column": 23,
                      "line": 4,
                    },
                  },
                  "range": [
                    138,
                    148,
                  ],
                  "type": "CallExpression",
                  "typeArguments": null,
                },
                "loc": {
                  "end": {
                    "column": 34,
                    "line": 4,
                  },
                  "source": null,
                  "start": {
                    "column": 16,
                    "line": 4,
                  },
                },
                "range": [
                  131,
                  149,
                ],
                "type": "ReturnStatement",
              },
            ],
            "loc": {
              "end": {
                "column": 13,
                "line": 5,
              },
              "source": null,
              "start": {
                "column": 57,
                "line": 3,
              },
            },
            "range": [
              113,
              163,
            ],
            "type": "BlockStatement",
          },
          "expression": false,
          "generator": false,
          "id": {
            "loc": {
              "end": {
                "column": 46,
                "line": 3,
              },
              "source": null,
              "start": {
                "column": 36,
                "line": 3,
              },
            },
            "name": "pathToFile",
            "optional": false,
            "range": [
              92,
              102,
            ],
            "type": "Identifier",
            "typeAnnotation": null,
          },
          "loc": {
            "end": {
              "column": 13,
              "line": 5,
            },
            "source": null,
            "start": {
              "column": 27,
              "line": 3,
            },
          },
          "params": [],
          "predicate": null,
          "range": [
            83,
            163,
          ],
          "returnType": {
            "loc": {
              "end": {
                "column": 56,
                "line": 3,
              },
              "source": null,
              "start": {
                "column": 48,
                "line": 3,
              },
            },
            "range": [
              104,
              112,
            ],
            "type": "TypeAnnotation",
            "typeAnnotation": {
              "loc": {
                "end": {
                  "column": 56,
                  "line": 3,
                },
                "source": null,
                "start": {
                  "column": 50,
                  "line": 3,
                },
              },
              "range": [
                106,
                112,
              ],
              "type": "StringTypeAnnotation",
            },
          },
          "type": "FunctionDeclaration",
          "typeParameters": null,
        },
        "exportKind": "value",
        "loc": {
          "end": {
            "column": 13,
            "line": 5,
          },
          "source": null,
          "start": {
            "column": 12,
            "line": 3,
          },
        },
        "range": [
          68,
          163,
        ],
        "type": "ExportDefaultDeclaration",
      },
    ],
    "comments": [
      {
        "loc": {
          "end": {
            "column": 8,
            "line": 1,
          },
          "source": null,
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "range": [
          0,
          8,
        ],
        "type": "Line",
        "value": " @flow",
      },
    ],
    "errors": [],
    "leadingComments": [
      {
        "loc": {
          "end": {
            "column": 8,
            "line": 1,
          },
          "source": null,
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "range": [
          0,
          8,
        ],
        "type": "Line",
        "value": " @flow",
      },
    ],
    "loc": {
      "end": {
        "column": 13,
        "line": 5,
      },
      "source": null,
      "start": {
        "column": 12,
        "line": 2,
      },
    },
    "range": [
      21,
      163,
    ],
    "type": "Program",
  },
  "type": "ast-jstree",
}
`;
